From 04af7e9cef36c78f18e575645edc4e34e912bce1 Mon Sep 17 00:00:00 2001
From: Ziad Elhanafy <ziad.elhanafy@arm.com>
Date: Thu, 21 Dec 2023 16:57:25 +0000
Subject: [PATCH] rss: Add common platform logger

platform_log defines log messages macros to be used by
the platform code It allows defining the module name
to be added at the beginning of the log message.
Based on build type PLAT_LOG_LEVEL is defined.
In case of Debug/RelWithDebInfo PLAT_LOG_LEVEL is defined
to Debug level else it is defined to OFF.

usage in source file:
...
INFO("msg");
ERROR("msg");
WARN("msg");
VERBOSE("msg");
DEBUG("msg");
...

Upstream-Status: Pending
Signed-off-by: Ziad Elhanafy <ziad.elhanafy@arm.com>
---
 .../ext/target/arm/rss/common/CMakeLists.txt  | 10 ++++
 .../ext/target/arm/rss/common/config.cmake    |  6 ++
 .../ext/target/arm/rss/common/platform_log.h  | 60 +++++++++++++++++++
 .../rss/common/provisioning/CMakeLists.txt    |  1 +
 .../provisioning/bundle_cm/CMakeLists.txt     |  5 ++
 5 files changed, 82 insertions(+)
 create mode 100644 platform/ext/target/arm/rss/common/platform_log.h

diff --git a/platform/ext/target/arm/rss/common/CMakeLists.txt b/platform/ext/target/arm/rss/common/CMakeLists.txt
index 48ee219ec..7e2a61ba9 100644
--- a/platform/ext/target/arm/rss/common/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/common/CMakeLists.txt
@@ -182,6 +182,7 @@ target_compile_definitions(platform_s
         ATTEST_KEY_BITS=${ATTEST_KEY_BITS}
         spm_memcpy=spm_dma_memcpy
     PRIVATE
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
         # Needed for DMA-350 library
         CMSIS_device_header="rss.h"
         $<$<BOOL:${PLATFORM_SVC_HANDLERS}>:PLATFORM_SVC_HANDLERS>
@@ -291,6 +292,7 @@ target_include_directories(platform_bl2
 
 target_compile_definitions(platform_bl2
     PRIVATE
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
         # Needed for DMA-350 library
         $<$<BOOL:${PLATFORM_HAS_BOOT_DMA}>:PLATFORM_HAS_BOOT_DMA>
         $<$<BOOL:${PLATFORM_BOOT_DMA_MIN_SIZE_REQ}>:BOOT_DMA_MIN_SIZE_REQ=${PLATFORM_BOOT_DMA_MIN_SIZE_REQ}>
@@ -367,6 +369,11 @@ target_link_libraries(cc312_rom
         platform_bl1_1
 )
 
+target_compile_definitions(platform_bl1_1
+    PRIVATE
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
+)
+
 #========================= Platform BL1_2 =====================================#
 
 target_sources(platform_bl1_2
@@ -393,6 +400,7 @@ target_compile_definitions(platform_bl1_2
         $<$<AND:$<BOOL:${CONFIG_TFM_BOOT_STORE_MEASUREMENTS}>,$<BOOL:${TFM_PARTITION_MEASURED_BOOT}>>:MEASURED_BOOT_API>
         $<$<BOOL:${RSS_AMOUNT}>:RSS_AMOUNT=${RSS_AMOUNT}>
         $<$<BOOL:${TFM_FWU_AGENT}>:TFM_FWU_AGENT>
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
 )
 
 if(RSS_USE_HOST_FLASH)
@@ -435,6 +443,7 @@ if(TFM_MULTI_CORE_TOPOLOGY)
     target_compile_definitions(tfm_psa_rot_partition_ns_agent_mailbox
         PRIVATE
             $<$<BOOL:${TFM_FWU_AGENT}>:TFM_FWU_AGENT>
+            PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
     )
 
 endif()
@@ -453,6 +462,7 @@ target_sources(tfm_spm
 target_compile_definitions(tfm_spm
     PRIVATE
         $<$<BOOL:${TFM_FWU_AGENT}>:TFM_FWU_AGENT>
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
 )
 
 #=========================== CPAK generator tool ===============================
diff --git a/platform/ext/target/arm/rss/common/config.cmake b/platform/ext/target/arm/rss/common/config.cmake
index 9886734e1..2f02162cf 100644
--- a/platform/ext/target/arm/rss/common/config.cmake
+++ b/platform/ext/target/arm/rss/common/config.cmake
@@ -86,3 +86,9 @@ set(PLAT_MHU_VERSION                    2          CACHE STRING   "Supported MHU
 set(RSS_AMOUNT                          1          CACHE STRING  "Amount of RSSes in the system")
 
 set(BL1_SHARED_SYMBOLS_PATH             ${CMAKE_CURRENT_LIST_DIR}/bl1/bl1_1_shared_symbols.txt CACHE FILEPATH "Path to list of symbols that BL1_1 that can be referenced from BL1_2")
+
+if (${CMAKE_BUILD_TYPE} STREQUAL Debug OR ${CMAKE_BUILD_TYPE} STREQUAL RelWithDebInfo)
+  set(PLAT_LOG_LEVEL                     4        CACHE STRING    "Set platform log level.")
+else()
+  set(PLAT_LOG_LEVEL                     0        CACHE STRING    "Set platform log level.")
+endif()
diff --git a/platform/ext/target/arm/rss/common/platform_log.h b/platform/ext/target/arm/rss/common/platform_log.h
new file mode 100644
index 000000000..b3a6e9802
--- /dev/null
+++ b/platform/ext/target/arm/rss/common/platform_log.h
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __PLATFORM_LOG_H__
+#define __PLATFORM_LOG_H__
+
+#define PLAT_LOG_LEVEL_OFF       (0)
+#define PLAT_LOG_LEVEL_ERROR     (1)
+#define PLAT_LOG_LEVEL_WARN      (2)
+#define PLAT_LOG_LEVEL_INFO      (3)
+#define PLAT_LOG_LEVEL_DEBUG     (4)
+
+#ifndef PLAT_LOG_MODULE_NAME
+#define MODULE_NAME_STR " "
+#else
+#define MODULE_NAME_STR "["PLAT_LOG_MODULE_NAME"]: "
+#endif
+
+#ifndef PLAT_LOG_LEVEL
+#warning "Logging level is not defined, default is PLAT_LOG_LEVEL_ERROR."
+#define PLAT_LOG_LEVEL   PLAT_LOG_LEVEL_ERROR
+#endif
+
+
+/* System can override PRINTF with other rich format function*/
+#ifndef PRINTF
+#if PLAT_LOG_LEVEL > PLAT_LOG_LEVEL_OFF
+#include <stdio.h>
+#define PRINTF printf
+#endif
+#endif
+
+#if PLAT_LOG_LEVEL >= PLAT_LOG_LEVEL_ERROR
+    #define ERROR(f_, ...) do { PRINTF("\033[31;4m[ERR]:\033[m%s"f_"\r\n", MODULE_NAME_STR, ##__VA_ARGS__);  } while (0)
+#else
+    #define ERROR(f_, ...) do {  } while(0)
+#endif
+#if PLAT_LOG_LEVEL >= PLAT_LOG_LEVEL_WARN
+    #define WARN(f_, ...) do { PRINTF("\033[33;4m[WRN]:\033[m%s"f_"\r\n", MODULE_NAME_STR, ##__VA_ARGS__);  } while (0)
+#else
+    #define WARN(f_, ...) do {  } while(0)
+#endif
+#if PLAT_LOG_LEVEL >= PLAT_LOG_LEVEL_INFO
+    #define INFO(f_, ...) do { PRINTF("[INF]:%s"f_"\r\n", MODULE_NAME_STR, ##__VA_ARGS__);  } while (0)
+#else
+    #define INFO(f_, ...) do {  } while(0)
+#endif
+#if PLAT_LOG_LEVEL >= PLAT_LOG_LEVEL_DEBUG
+    #define VERBOSE(f_, ...) do { PRINTF("[DBG]:%s" f_"\r\n",MODULE_NAME_STR, ##__VA_ARGS__);  } while (0)
+    #define DEBUG(f_, ...) do { PRINTF("[DBG]:%s" f_"\r\n",MODULE_NAME_STR, ##__VA_ARGS__);  } while (0)
+#else
+    #define VERBOSE(f_, ...) do {  } while(0)
+    #define DEBUG(f_, ...) do {  } while(0)
+#endif
+
+#endif /* __PLATFORM_LOG_H__ */
diff --git a/platform/ext/target/arm/rss/common/provisioning/CMakeLists.txt b/platform/ext/target/arm/rss/common/provisioning/CMakeLists.txt
index 5e0af35cb..c2c9afa75 100644
--- a/platform/ext/target/arm/rss/common/provisioning/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/common/provisioning/CMakeLists.txt
@@ -32,4 +32,5 @@ target_include_directories(bl1_1_lib
 target_compile_definitions(bl1_1_lib
     INTERFACE
         $<$<BOOL:${TFM_FWU_AGENT}>:TFM_FWU_AGENT>
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
 )
diff --git a/platform/ext/target/arm/rss/common/provisioning/bundle_cm/CMakeLists.txt b/platform/ext/target/arm/rss/common/provisioning/bundle_cm/CMakeLists.txt
index 74cc24c60..a498cde0f 100644
--- a/platform/ext/target/arm/rss/common/provisioning/bundle_cm/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/common/provisioning/bundle_cm/CMakeLists.txt
@@ -38,6 +38,11 @@ target_sources(cm_provisioning_bundle
         $<$<BOOL:${TFM_FWU_AGENT}>:${TARGET_PLATFORM_PATH}/fw_update_agent/fwu_agent.c>
 )
 
+target_compile_definitions(cm_provisioning_bundle
+    PRIVATE
+        PLAT_LOG_LEVEL=${PLAT_LOG_LEVEL}
+)
+
 target_include_directories(cm_provisioning_bundle
     PRIVATE
         ../bundle_common
-- 
2.25.1

